go-controller/pkg/clustermanager/healthcheck/controller.go

```diff
@@ -0,0 +1,426 @@
+package healthcheck
+
+import (
+	"context"
+	"fmt"
+	"math/rand"
+	"net"
+	"sync"
+	"time"
+
+	"github.com/google/go-cmp/cmp"
+	"github.com/google/go-cmp/cmp/cmpopts"
+
+	controllerutil "github.com/ovn-org/ovn-kubernetes/go-controller/pkg/controller"
+	"github.com/ovn-org/ovn-kubernetes/go-controller/pkg/ovn/healthcheck"
+	"github.com/ovn-org/ovn-kubernetes/go-controller/pkg/types"
+	"github.com/ovn-org/ovn-kubernetes/go-controller/pkg/util"
+	"github.com/ovn-org/ovn-kubernetes/go-controller/pkg/util/timeoutqueue"
+
+	corev1 "k8s.io/api/core/v1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	coreinformers "k8s.io/client-go/informers/core/v1"
+	"k8s.io/client-go/util/workqueue"
+	"k8s.io/klog/v2"
+)
+
+const (
+	// period in seconds of each node probe
+	period = 1
+	// jitter in milliseconds added to the period to distribute probes over
+	jitter = 200
+)
+
+// nodeState contians the latest declared health state for a node and the latest
+// time it was available
+type nodeState struct {
+	health    HealthState
+	available time.Time
+}
+
+// transitionNodeState controls the transition to a new node state from the
+// latest node state and an actual health state, return a string reasoning about
+// the state change
+func transitionNodeState(old nodeState, health HealthState, on time.Time, timeout time.Duration, reset bool) (new nodeState, reason string) {
+	new = old
+	reason = "health state changed"
+
+	switch health {
+	case AVAILABLE:
+		new.health = AVAILABLE
+		new.available = on
+	case UNREACHABLE:
+		// only set UNREACHABLE after timeout
+		if time.Since(old.available) > timeout {
+			new.health = UNREACHABLE
+			reason = fmt.Sprintf("not available since %s", old.available)
+		}
+	default:
+		new.health = health
+	}
+
+	if reset {
+		reason = "health state reset"
+		new.health = health
+	}
+
+	return
+}
+
+// nodeInfo contains the IPs of the node and a function to cancel the current
+// probing task for the node
+type nodeInfo struct {
+	ips        []net.IP
+	cancelTask context.CancelFunc
+}
+
+// probeTask represents a probing task for a node including the the context it
+// is executed under and the client used for the probe
+type probeTask struct {
+	node   string
+	client healthcheck.EgressIPHealthClient
+	ctx    context.Context
+}
+
+func (nt probeTask) cancelled() bool {
+	select {
+	case <-nt.ctx.Done():
+		return true
+	default:
+	}
+	return false
+}
+
+type controller struct {
+	sync.RWMutex
+
+	ctx context.Context
+
+	controller   controllerutil.Reconciler
```

Comment:
having controller hold another thing called controller is awkward. Maybe consider renaming it to something more specific